app/src/main/java/com/zulip/android/networking/HTTPRequest.java

```diff
@@ -1,141 +1,129 @@
 
 /**
  * Simplified HTTP request API
  */
 public class HTTPRequest {
-    ZulipApp app;
-    List<NameValuePair> nameValuePairs;
-    HttpRequestBase request;
-    HttpResponse response;
+    private ZulipApp app;
     volatile boolean aborting = false;
+    private HashMap<String, String> properties;
+    private OkHttpClient okHttpClient;
+    private Response response = null;
+    private Request request;
+    private String method, path;
+
+    public void setMethodAndUrl(String method, String URL) {
+        this.method = method;
+        this.path = URL;
+    }
 
     public HTTPRequest(ZulipApp app) {
-        nameValuePairs = new ArrayList<NameValuePair>();
+        properties = new HashMap<>();
         this.app = app;
+        okHttpClient = new OkHttpClient.Builder().readTimeout(60, TimeUnit.SECONDS).build();
     }
 
+
     public void setProperty(String key, String value) {
-        this.nameValuePairs.add(new BasicNameValuePair(key, value));
+        properties.put((key == null) ? "" : key, (value == null) ? "" : value);
     }
 
-    public void clearProperties() {
-        this.nameValuePairs.clear();
+    void clearProperties() {
+        properties.clear();
     }
 
-    public void abort() {
+    void abort() {
         aborting = true;
-        if (request != null) {
+        if (response != null) {
             (new AsyncTask<Void, Void, Void>() {
                 @Override
                 protected Void doInBackground(Void... voids) {
-                    request.abort();
+                    response.body().close();
                     return null;
                 }
             }).execute();
         }
     }
 
-    // Java doesn't support a request body on DELETE requests, but RFC2616 does
-    // not prohibit it.
-    class HttpDeleteWithReq extends HttpPost {
-        HttpDeleteWithReq(String url) {
-            super(url);
+    public Response execute() throws IOException {
+        if (method == null)
+            throw new IOException(app.getString(R.string.method_null));
+        Request.Builder requestBuilder = new Request.Builder();
+        String url = app.getServerURI() + path;
+        requestBuilder.addHeader("client", "Android");
+
+        switch (method) {
+            case "GET":
+                requestBuilder.url(generateURL(url, properties)).get();
+                break;
+            case "DELETE":
+                FormBody.Builder formDeleteBody = new FormBody.Builder();
+                for (Map.Entry<String, String> map : properties.entrySet()) {
+                    formDeleteBody.addEncoded(map.getKey(), map.getValue());
+                }
+                requestBuilder.url(url).delete(formDeleteBody.build());
+                break;
+            case "POST":
+                FormBody.Builder formBody = new FormBody.Builder();
+                for (Map.Entry<String, String> map : properties.entrySet()) {
+                    formBody.addEncoded(map.getKey(), map.getValue());
```

Comment:
I'm running into a problem when logging in with an email.  This API `addEncoded` is transforming the `@` into the encoded version `%40`, making it such that I can't log in. :(